[InstCombine] Try to fold add into GEP x, C

davemgreen opened the pull request:
This attempts to GEP p, (x + C1), C2 with GEP p, x, C2+C1*S, removing the need
for the add by folding it into the following constant offset in the GEP. It
drops inbounds from geps, as they may no longer be valid.

Example proof using a much reduced pointer address space:
https://alive2.llvm.org/ce/z/hG8xfF

davemgreen committed:
[InstCombine] Add Gep + Add tests. NFC

dtcxzyw commented on llvm/lib/Transforms/InstCombine/InstructionCombining.cpp:
```diff
@@ -2893,6 +2893,37 @@ Instruction *InstCombinerImpl::visitGetElementPtrInst(GetElementPtrInst &GEP) {
+          *Next = ConstantInt::get((*Next)->getType(),
+                                    *C2 + *C1 * (Scale1 / Scale2));
+          addToWorklist(Add);
+          return &GEP;
```
Is it safe to propagate `inbounds` flag?
Alive2: https://alive2.llvm.org/ce/z/QD6chN

nikic reviewed (changes requested):
I'm not really willing to take patches for the formation of "big GEPs" anymore at this point. Our canonicalization direction should start to go in the opposite direction, breaking up these GEPs into smaller parts. At that point, the existing gep+add to gep+gep canonicalization will also handle this.

davemgreen reviewed (commented):
> I'm not really willing to take patches for the formation of "big GEPs" anymore at this point. Our canonicalization direction should start to go in the opposite direction, breaking up these GEPs into smaller parts. At that point, the existing gep+add to gep+gep canonicalization will also handle this.

Yeah I thought you might bring that up :)  I was hoping we could be reasonable and take optimizations like this as a short-term improvement until the canonicalisation is ready, if the patch was simple enough. (This wasn't attempting to make bigger geps, just change the ones that already existed). I had seen that with smaller geps it was already handled better, this just reduces some of the differences between the current larger geps and the future smaller ones.

davemgreen committed:
[InstCombine] Try to fold add into GEP

This attempts to GEP p, (x + C1), C2 with GEP p, x, C2+C1*S, removing the need
for the add by folding it into the following constant offset in the GEP. It
does not attempt to alter inbounds geps, to avoid the need to drop the
inbounds.

Example proof using a much reduced pointer address space:
https://alive2.llvm.org/ce/z/hG8xfF

davemgreen commented:
I'd like to try this again if I can. This doesn't seem like an unreasonable patch to take in the sort-term, and as far as I understand is a (little) step towards where we expect codegen to end up. As in - it performs an optimization that will happen if geps are canonicalized to individual offsets, bringing the final codegen closer to what it will later be. And doesn't create larger geps, just transforms existing large geps to be a little different. I've given the patch a rebase. Let me know what you think.

nikic commented:
I think we should at least try to solve this by breaking up the GEPs first. If that is not feasible without a lot of effort right now (because too many other opts depend on the status quo), then we can consider this direction.

davemgreen commented:
> bringing the final codegen closer to what it will later be

Unfortunately it looks like this might not be exactly true, and canonicalizing the geps (at least in the way I have tried it) does not produce the same final output I was hoping for, I think partly because of multiple uses.

nikic commented:
> > bringing the final codegen closer to what it will later be
> 
> Unfortunately it looks like this might not be exactly true, and canonicalizing the geps (at least in the way I have tried it) does not produce the same final output I was hoping for, I think partly because of multiple uses.

Can you share the motivating example?
